Add support to install master branch when using WooCommerce Beta Tester

samueljseay opened the pull request:
## Changes

* Fix a bug where the admin menu wasn't being registered early enough, causing it to not be added to the navigatable pages
* Add support for installing the trunk branch in the UI and add a little more error handling for failing to fetch manifest
* Add support to WP CLI for installing the trunk branch

## Testing Instructions

1. Checkout this branch, mount the beta tester plugin in your local dev environment. Also have WooCommerce installed and activated.
2. Ensure composer has been run (it happens on pnpm postinstall)
3. Ensure JS assets are built. (In dev fastest way is just run `pnpm start` from the beta tester directory)
4. Now if you navigate to the live branches screen in wp-admin via `WooCommerce > Live Branches` you should see that trunk is listed as an available branch to install in the dropdown. Try "Install and Activate". It should be installed and activated as the current version
5. Now ideally reset your environment to test WP-CLI (let me know if you need assistance with using wp-cli in your env, its different for every env setup).  Run `wp wc-beta-tester install trunk` and then `wp wc-beta-tester deactivate_woocommerce` and then `wp wc-beta-tester activate trunk`. These commands should all succeed then you can go to the live branches menu in wp-admin as in step 4, and see that trunk is installed and activated.

samueljseay committed:
Pin linter versions in syncpack

samueljseay committed:
Fix version mismatches with syncpack and update lock file.

samueljseay committed:
Stop auto install peers from installing wrong prettier version.

samueljseay committed:
Fix a bug where live branches JS can be registered too late.

samueljseay committed:
Include trunk as installable branch in live branches

samueljseay committed:
Add changelog entry

samueljseay committed:
Add support for getting trunk branch when using CLI

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/trunk-to-beta-tester

lsinger commented on plugins/woocommerce-beta-tester/api/live-branches/manifest.php:
```diff
@@ -29,5 +29,7 @@ function fetch_live_branches_manifest() {
 		$value->install_status = $installer->check_install_status( $value->version );
 	}
 
+	$obj->master->install_status = $installer->check_install_status( $obj->master->version );
```
How do we handle errors here (and in the upper part of this function)? What if there's no `master` property or it doesn't have an `install_status` or the whole JSON is malformed -- do we handle these things outside of this function somehow? 🤔

samueljseay replied:
No we don't really handle errors here if the manifest is not retrieved. I'll see if I can improve this. Thanks for calling it out!

samueljseay replied:
@lsinger I made a couple changes here: https://github.com/woocommerce/woocommerce/pull/38536/commits/ef084342f1cce22665d8269f7252f310aa921a30

1. Handle a variety of error cases: error in response, error in json decode, props not on object. In those cases return 500 response

2. I noticed that triggering an error response was causing a crash in the JS app, I have fixed that by only rendering the branch list if there is no error.

You can test one of the error cases by changing the URL at [this line](https://github.com/woocommerce/woocommerce/blob/ef084342f1cce22665d8269f7252f310aa921a30/plugins/woocommerce-beta-tester/api/live-branches/manifest.php#L22) to something incorrect. On the front-end you'll get the generic "Something went wrong" message. (Make sure you rebuild the JS of course!)

lsinger commented on plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-live-branches.php:
```diff
@@ -15,10 +15,8 @@ class WC_Beta_Tester_Live_Branches {
 	public function __construct() {
-		add_action( 'admin_enqueue_scripts', array( $this, 'register_scripts' ) );
-
-		// By the time this code runs it appears too late to hook into `admin_menu`.
```
This comment seems to call out specifically that hooking into `admin_menu` won't work ... and then you do it! 😅 Was the comment inaccurate? Did something else change? Just trying to understand. 🤔

samueljseay replied:
Yep you'll notice I've changed where this code is loaded to load much earlier so I can hook into this. I require this file now in the main plugin file to ensure that it can hook earlier

samueljseay replied:
@lsinger loaded earlier here https://github.com/woocommerce/woocommerce/pull/38536/files#diff-48c214be161d3e5f0d580e862544417631022d0cf561892791583261cc65707dR137

lsinger commented on plugins/woocommerce-beta-tester/src/live-branches/hooks/live-branches.tsx:
```diff
@@ -16,31 +16,44 @@ export type Branch = {
+					}
+				) as Branch[];
+
+				console.log( res.master, prBranches );
```
The `console.log` should probably go?

samueljseay replied:
Whoops! Thanks for spotting this!

samueljseay replied:
Resolved https://github.com/woocommerce/woocommerce/pull/38536/commits/1d75ec22b7bf0c714143a561afed8862a65fa9bd

samueljseay committed:
Remove accidentally added console log statement.

samueljseay committed:
Improve error handling in manifest retrieval, ensure that JS app doesn't throw

samueljseay commented:
Thanks for the great review @lsinger its ready for another look. 🙇🏻

samueljseay committed:
PHPCS linting.

lsinger commented:
@samueljseay thanks for making those changes! Unfortunately I have a few additional questions. 

I used wp-env from the beta-tester directory. Not having WooCommerce installed yet on this initial run, I ran into a fatal because `woocommerce-beta-tester.php` relies on a constant `WC_ADMIN_APP` being defined. I added a `defined` check around that, which made my wp-env sits work again and let me install WooCommerce now. 

1. Would you mind adding that check to this PR? (just wrap line 91 in `if ( defined( 'WC_ADMIN_APP' ) ) { ...}`)

I then got the following warning on multiple pages -- e.g., the Orders page and the plugins page:

> Notice: Function get_current_page was called incorrectly. Current page retrieval should be called on or after the `current_screen` hook. Please see [Debugging in WordPress](https://wordpress.org/documentation/article/debugging-in-wordpress/) for more information. (This message was added in version 0.16.0.) in /var/www/html/wp-includes/functions.php on line 5865

2. Is this possibly related to this PR? It goes away if I turn of beta-tester. 

When I (successfully) install trunk -- is there a way to get back off of live branches and on a version from the beta-tester "switch version" feature? When I try that I get a fatal because it's basically trying to have Woo installed twice -- once in the live-branches directory and once as the proper plugin that beta-tester works with (AFAIK). 

3. Is there a way to switch back, and is this behavior OK for now? Is there an easy change maybe that can help people not run into that fatal, maybe?

samueljseay commented:
@lsinger 🤔 

1. I think perhaps we should throw an error instead. beta tester still relies on WooCommerce as a dependency, if we just put the check around the registering of those scripts then the plugin will not function in other ways its intended to. How about we just raise a bug for this and decide on a path forward? it wasn't introduced by this PR. Also apologies, I probably missed "install WooCommerce" as part of the test instructions, When coming from Jurassic Ninja it installs WooCommerce by default alongside the beta tester.

2. I think its very unlikely introduced by this PR, that method is not called by this PR directly. Its a page controller method that sounds like its being called at wrong time in lifecycle for "some reason". Either way I haven't seen it locally, best we can do is if you have the full stack trace we can work out where the call originates from? If its code modified by this PR we can try address it, otherwise it sounds like we should raise another bug if we can reliably reproduce it.

3. Ah interesting. Yeah I went to great lengths to avoid double install of WC inside live branches, but I didn't account for trying to switch between live branch and the ones managed by live branches. Let's raise a bug to fix separately, its also not introduced by this PR and would be out of scope of this piece of work. But we should really address it soon, thank you for testing that.

lsinger commented:
> @lsinger 🤔
> 
> 1. I think perhaps we should throw an error instead. beta tester still relies on WooCommerce as a dependency, if we just put the check around the registering of those scripts then the plugin will not function in other ways its intended to. How about we just raise a bug for this and decide on a path forward? it wasn't introduced by this PR. Also apologies, I probably missed "install WooCommerce" as part of the test instructions, When coming from JN it installs WooCommerce by default alongside the beta tester.

Oh, when we add the check as I described it already gives a nice error message. ☺️ But agreed, should go in a separate issue: #38578

> 2. I think it's very unlikely introduced by this PR, that method is not called by this PR directly. It's a page controller method that sounds like its being called at wrong time in lifecycle for "some reason". Either way I haven't seen it locally, best we can do is if you have the full stack trace we can work out where the call originates from? If its code modified by this PR we can try address it, otherwise it sounds like we should raise another bug if we can reliably reproduce it.

I can reliably trigger it by switching from trunk (where it's absent) to this PR's branch. It goes away when I remove `add_action( 'admin_init', array( $this, 'register_scripts' ) );` -- I assume somewhere down the WordPress hooks rabbit hole this indeed does call that function. 

I played around with replacing `admin_init` with `admin_enqueue_scripts`, which according to the Codex seems to be the canonical way for registering and enqueueing scripts, but I then get an error telling me "" when I try to access the Live Branches page. I dug into this some more what wasn't able to find the right way to do it, unfortunately. I have a gut feeling that this all could be rooted in the `wc_admin_register_page` call, which might require a different hook strategy than standard WordPress, but I haven't confirmed this. 

> 3. Ah interesting. Yeah I went to great lengths to avoid double install of WC inside live branches, but I didn't account for trying to switch between live branch and the ones managed by older beta tester features. Let's raise a bug to fix separately, it would be out of scope of this piece of work. But we should really address it soon, thank you for testing that. (Probably the fix will be exploring how we can deactivate any existing WooCommerce versions that don't have the "live branches" prefix, but I think it can also be done as follow up since this PR doesn't introduce this issue).

Additional issue sounds good, thanks for the reminder to keep PRs focused. ☺️ #38579

samueljseay committed:
Register scripts later, put them in head not footer.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/trunk-to-beta-tester

samueljseay committed:
Dont use WC objects to determine admin.

samueljseay committed:
Go back to admin_init so that our JS runs before hooks are run.

samueljseay commented:
@lsinger 

> Oh, when we add the check as I described it already gives a nice error message

Can you explain what you mean here? How will not registering front-end scripts give a nice error message? 

> I played around with replacing admin_init with admin_enqueue_scripts, which according to the Codex seems to be the canonical way for registering and enqueueing scripts, 

I dug into this and what I found is:

1. If I use admin_enqueue_scripts it won't work right now. That's because the [hook that I use ](https://github.com/woocommerce/woocommerce/blob/fc96fa739c8be7457cc30b81617c5d9d3d3324b9/plugins/woocommerce-beta-tester/src/live-branches/index.tsx#L11) has already been run before my JS runs. I think this is possibly a bug with WooCommerce, I've added a task to my list to try produce a minimal reproduction of this because I think it seems like the hooks are running much too early. All our examples suggest registering scripts in admin_enqueue, but I wonder how that can work if our script is loaded after the wc one that runs hooks. I can't find any doco on this problem.

2. Another thing I tried was going back to admin_enqueue but registering my script in the `head` instead of footer. But that throws an error in WC scripts that the global wcAdmin object is not available. I guess this is because it loads it too early in the `head` before the script has run that adds that global. 

None of this has anything to do with how I've approached things in this plugin, these are flaws with how script ordering is done and how we have to use a JS hook to add the page to WooCommerce navigation. We could maybe just not use wc admin page registration in future to avoid this problem.

> I can reliably trigger it by switching from trunk (where it's absent) to this PR's branch. It goes away when I remove add_action( 'admin_init', array( $this, 'register_scripts' ) ); -- I assume somewhere down the WordPress hooks rabbit hole this indeed does call that function.

Sure, but, since I couldn't repro **I really need the stack trace from your code to determine where the origin of the PHP error is.**

That said though, I have taken a guess, since the only place I call page controller code is to determine if its admin, and I really only need `is_admin` from core Wordpress, I have changed it to use that instead. It's a guess but could you see if now that I removed ref to PageController objects if it fixes the problem you saw?

For now we must stick with `admin_init` for our JS hook to be called and I'll investigate if I can repro in a basic plugin.

lsinger reviewed (approved):
Thank you @samueljseay for the changes -- things work well and as intended now.

lsinger commented:
> Can you explain what you mean here? How will not registering front-end scripts give a nice error message? I'm sure you're right but I think I just don't understand what you're saying :)

@samueljseay not using the undefined WC_APP constant will keep the site from crashing, which will give the error message a chance to appear. I added some more details to #38578.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/trunk-to-beta-tester
